### PR TITLE
[IR] Introduce `llvm.allow.{runtime,ubsan}.check()`

### PR DESCRIPTION
**Pull request by vitalybuka**
The goal is to have ability to change logic compile time based on PGO data.

Our primary application is removing UBSAN checks from hot code.
Then we'd like to use this for libc++ hardening and regular debug asserts.
Previous attempt is #84214.

Benefits from special intrinsic vs #84214:
1. Resulting binary is 3% faster than removing traps (on "test-suite/MultiSource/Benchmarks" with PGO+ThinLTO)
2. Intrinsic can be used from source code to change behavior from C/C++ program. E.g. enabling asserts in cold code.
3. Easier to match basic blocks.

RFC: https://discourse.llvm.org/t/rfc-add-llvm-experimental-hot-intrinsic-or-llvm-hot/77641

### CODE REVIEW

**Comment by jroelofs** [comment]
Neat idea!

**Review by kstoimenov — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7281]: Missing globalisel and LowerIntrinsics handling

Comment (aeubanks) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7281]: we shouldn't even get to isel, this should be unconditionally lowered in the optimization pipeline and any instance of this intrinsic getting to isel should be considered a bug (e.g. coroutines, LowerMatrixIntrinsicsPass)

**Comment by aeubanks** [comment]
+1, this is a cool addition!

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp
Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7281]: The optimization pipeline isn't mandatory, -O0 must still work

Comment (aeubanks) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7281]: you are still required to run the -O0 "optimization" pipeline to do things like lowering (e.g. the examples I gave above) before codegen

File: llvm/include/llvm/IR/Intrinsics.td
Comment (dvyukov) [inline llvm/include/llvm/IR/Intrinsics.td:1727]: Can't IntrWriteMem have significant effect on performance of the generated code? Why exactly do we need it? A comment would be useful. It's not writing to memory, is there a more precise attribute to capture what we need?

File: llvm/docs/LangRef.rst
Comment (vitalybuka) [reply llvm/docs/LangRef.rst:27684]: copypasted mistake from @llvm.experimental.widenable.condition :)

File: llvm/include/llvm/IR/Intrinsics.td
Comment (vitalybuka) [reply llvm/include/llvm/IR/Intrinsics.td:1727]: No, we have no more precise attribute. E.g. int_sideeffect is declared this way as well.
There are a few comments for similar cases around, I'll add one here to.
We need to lock intrinsic in place, without it, it will be moved to the entry block and value shared between users. So it will work as function level, vs block level.

With ThinLTO it should not be a problem, in postlink we rerun simplification and optimization with intrinsic lowered to  true or false. I tried to remove this attribute and make it function level, but can't measure perf difference.

Alternative is to have intrinsic without sideeffects but with dependency like this:
```
%condition = call i1 @llvm.experimenta.hot(i1 %ubsan_condition)
```
This way we can lock it in place. But this will not help us with assert() with sideeffects as we discussed before.

So I tried both with UBSAN, ThinLTO and plain, and 

```
call i1 @llvm.experimenta.hot()
```

is consistently, but close to noise, faster then:

```
call i1 @llvm.experimenta.hot(i1 %ubsan_condition)
```

I believe because  `or i1 %ubsan, @llvm.experimenta.hot()`  is more predictable for early transformation passes than `@llvm.experimenta.hot(i1 %ubsan_condition)`.

Both intrinsic approaches faster then removing traps.

**Comment by vitalybuka** [comment]
> Please submit an RFC on discourse for this change.

https://discourse.llvm.org/t/rfc-add-llvm-experimental-hot-intrinsic-or-llvm-hot/77641

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp
Comment (vitalybuka) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:7281]: I guess we sill want some default lowering, e.g. for `-opt-bisect-limit=` working when we stop pipeline before the PGO based pass.

Comment (vitalybuka) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp]: moving out lowering into a separate CL

File: llvm/docs/LangRef.rst
Comment (arsenm) [inline llvm/docs/LangRef.rst:27739]:
```suggestion
If intrinsic is evaluated as ``true``, program must check ubsan condition, and
```

File: llvm/docs/LangRef.rst
Comment (arsenm) [inline llvm/docs/LangRef.rst:27740]:
```suggestion
report if needed. If intrinsic is evaluated as ``false``, program must 
```

File: llvm/docs/LangRef.rst
Comment (arsenm) [inline llvm/docs/LangRef.rst:27802]:
```suggestion
If intrinsic is evaluated as ``true``, program should execute a guarded checks.
If intrinsic is evaluated as ``false``, the program shoud avoid any unnececary
```

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27700]: Fix underline length :)

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27718]:
```suggestion
This intrinsic is the ubsan specific version of ``@llvm.allow.runtime.check()``.
```
or "an"

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27725]: Leftover.

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27752]:
```suggestion
  cont:
```
etc

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27748]:
```suggestion
    %not.allow = xor i1 %allow, true
```

File: llvm/docs/LangRef.rst
Comment (vitalybuka) [reply llvm/docs/LangRef.rst:27739]: That's getting embracing. Fixing my spell checker.

**Comment by vitalybuka** [comment]
How can we move forward here?

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27788]: Wrong intrinsic

File: llvm/docs/LangRef.rst
Comment (nikic) [inline llvm/docs/LangRef.rst:27798]:
```suggestion
check. If the intrinsic is evaluated as ``false``, the program should avoid any
```

Comment (nikic) [reply llvm/docs/LangRef.rst:27798]: (Or drop the "a".)

**Review by nikic — APPROVED** [review]
LGTM

I might have used allow -> should in the name, but this looks ok as well.

**Comment by vitalybuka** [comment]
@efriedma-quic @arsenm Would you like me to wait for your approval?

File: llvm/docs/LangRef.rst
Comment (MaskRay) [inline llvm/docs/LangRef.rst:27743]: The two paragraphs can be rephrased like the following

When used in a branch condition, it selects one of the two paths:

- true: Executes the UBSan check and reports any issues detected.
- false: Skips the check, assuming its success to potentially remove the check.

**Review by MaskRay — APPROVED** [review]

**Comment by arsenm** [comment]
> @efriedma-quic @arsenm Would you like me to wait for your approval?

No

**Review by efriedma-quic — COMMENTED** [review]
Didn't review carefully, but looks fine.

**Pull request merged**
